res/controllers/Denon-MC7000-scripts.js

```diff
@@ -529,14 +532,20 @@ MC7000.PadButtons = function(channel, control, value, status, group) {
             }
         }
     } else if (MC7000.PADModeRoll[deckNumber]) {
-        // TODO(all): check for actual beatloop_size and apply back after a PAD Roll
         i = control - 0x14;
-        if (control === 0x14 + i && value > 0x00) {
+        if (value > 0x00) {
+            // Save current loop size, then activate roll loop
+            MC7000.beatLoopSizeBeforeRoll = engine.getValue(group, "beatloop_size");
```

Comment:
As a small caveat, if a user holds multiple roll buttons in (overlapping) succession, the last roll size will be restored (this is the simplest implementation, but perhaps not necessarily the most intuitive behavior). Consider the following scenario:

- Initial beatloop size is 4
- User holds 1/16 roll button
  - Beatloop size of 4 is saved
  - Beatloop size is set to 1/16
- User holds 1/8 roll button
  - Beatloop size of 1/16 is saved (overwriting the saved size of 4)
  - Beatloop size is set to 1/8
- User releases 1/8 roll button
  - Beatloop size is restored to 1/16
- User releases 1/16 roll button
  - Beatloop size does not change (since the loop size was already restored)

Perhaps we should check if `beatLoopSizeBeforeRoll` is non-undefined to avoid overwriting it? Still, there could be scenarios where this logic is thrown off, in particular when any loop size is restored before the user releases all roll buttons.

Intuitively it feels like the beatloop size of 4 should be restored (i.e. the beatloop size before _any_ roll button was pressed).

The best solution would therefore probably be to keep track of a sort of 'button-down-count' that first restores the loop when back to zero (i.e. all roll buttons are released), but I am not sure whether this would be worth the effort and potential inconsistencies if some MIDI message where to get dropped for some reason.